python/src/aiconfig/ChatCompletion.py

```diff
@@ -199,3 +203,135 @@ async def run_and_await_serialize():
+            response, openai.Stream
+        )
+
+        # Convert Response to output for last prompt
+        if not stream:
+            outputs = extract_outputs_from_response(response)
+
+            # Add outputs to last prompt
+            serialized_prompts[-1].outputs = outputs
+
+            validate_and_add_prompts_to_config(serialized_prompts, output_aiconfig)
+
+            # Save config to file
+            output_aiconfig.save(output_config_file_path, include_outputs=True)
+
+            # Return original response
+            return response
+        else:
+            # If response is a stream, build the output as the stream iterated through. do_logic() becomes a generator.
+
+            # TODO: type
+            def generate_streamed_response() -> Generator[Any, None, None]:
+                stream_outputs = {}
+                messages = {}
+                for chunk in response:
+                    chunk_dict = chunk.model_dump(exclude_none=True)  # type: ignore [fixme]
+
+                    # streaming only returns one chunk, one choice at a time. The order in which the choices are returned is not guaranteed.
+                    messages = multi_choice_message_reducer(messages, chunk_dict)
+
+                    for choice in chunk_dict["choices"]:
+                        index = choice.get("index")
+                        accumulated_message_for_choice = messages.get(index, {})
+                        output = ExecuteResult(
+                            output_type="execute_result",
+                            data=copy.deepcopy(accumulated_message_for_choice),
+                            execution_count=index,
+                            metadata={"finish_reason": choice.get("finish_reason")},
+                        )
+                        stream_outputs[index] = output
+                    yield chunk
+                stream_outputs = [
+                    stream_outputs[i] for i in sorted(list(stream_outputs.keys()))
+                ]
+
+                # Add outputs to last prompt
+                serialized_prompts[-1].outputs = stream_outputs
+
+                validate_and_add_prompts_to_config(serialized_prompts, output_aiconfig)
+
+                # Save config to file
+                output_aiconfig.save(output_config_file_path, include_outputs=True)
+
+            return generate_streamed_response()
+
+    return _create_chat_completion_with_config_saving
+
+
+def get_completion_create_wrapped_openai(
+    output_aiconfig_ref: str | AIConfigRuntime,
+) -> ModuleType:
+    api = openai
+    new_module = cu.make_wrap_object(
+        api,
+        "chat.completions.create",
+        create_and_save_to_config_v2(
+            output_aiconfig_ref=output_aiconfig_ref, openai_api=api
+        ),
+    )
+    return cast(ModuleType, new_module)
+
+
+def get_completion_create_wrapped_openai_client(
+    output_aiconfig_ref: str | AIConfigRuntime,
+) -> openai.OpenAI:
+    api = openai.Client()
```

Comment:
When the user is instantiating a client, we can expect them to construct it with their own settings and options. For example, [instantiating an azure client](https://learn.microsoft.com/en-us/azure/ai-services/openai/quickstart?tabs=command-line%2Cpython-new&pivots=programming-language-python#:~:text=client%20%3D%20AzureOpenAI(%0A%20%20%20%20api_key%3Dos.getenv(%22AZURE_OPENAI_KEY%22)%2C%20%20%0A%20%20%20%20api_version%3D%222023%2D10%2D01%2Dpreview%22%2C%0A%20%20%20%20azure_endpoint%20%3D%20os.getenv(%22AZURE_OPENAI_ENDPOINT%22)%0A%20%20%20%20))

This should probably still work if you can pass in the client or something similar